extensions/sql-database-projects/src/test/projectController.test.ts

```diff
@@ -332,165 +336,171 @@ describe('ProjectsController: import operations', function (): void {
-
-		const projController = new ProjectsController(testContext.apiWrapper.object, new SqlDatabaseProjectTreeViewProvider());
-		await testUtils.shouldThrowSpecificError(async () => await projController.addDatabaseReference(new Project('FakePath')), constants.databaseNameRequired);
+		sinon.stub(vscode.window, 'showInputBox').returns(Promise.resolve(undefined));
+		sinon.stub(vscode.window, 'showQuickPick').returns(Promise.resolve({ label: constants.dacpac }));
+		sinon.stub(vscode.window, 'showOpenDialog').returns(Promise.resolve([vscode.Uri.file('FakePath')]));
+		const spy = sinon.spy(vscode.window, 'showErrorMessage');
+
+		const projController = new ProjectsController(new SqlDatabaseProjectTreeViewProvider());
+		await projController.addDatabaseReference(new Project('FakePath'));
+		should(spy.calledOnce).be.true('showErrorMessage should have been called');
 	});
 
 	it('Should return the correct system database', async function (): Promise<void> {
-		const projController = new ProjectsController(testContext.apiWrapper.object, new SqlDatabaseProjectTreeViewProvider());
+		const projController = new ProjectsController(new SqlDatabaseProjectTreeViewProvider());
 		const projFilePath = await testUtils.createTestSqlProjFile(baselines.openProjectFileBaseline);
 		const project: Project = new Project(projFilePath);
 		await project.readProjFile();
 
-		testContext.apiWrapper.setup(x => x.showQuickPick(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve({ label: constants.master }));
+		const stub = sinon.stub(vscode.window, 'showQuickPick').returns(Promise.resolve({ label: constants.master }));
 		let systemDb = await projController.getSystemDatabaseName(project);
 		should.equal(systemDb, SystemDatabase.master);
 
-		testContext.apiWrapper.setup(x => x.showQuickPick(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve({ label: constants.msdb }));
+		stub.returns(Promise.resolve({ label: constants.msdb }));
 		systemDb = await projController.getSystemDatabaseName(project);
 		should.equal(systemDb, SystemDatabase.msdb);
 
-		testContext.apiWrapper.setup(x => x.showQuickPick(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve(undefined));
+		stub.returns(Promise.resolve(undefined));
 		await testUtils.shouldThrowSpecificError(async () => await projController.getSystemDatabaseName(project), constants.systemDatabaseReferenceRequired);
 	});
 });
 
 describe('ProjectsController: round trip feature with SSDT', function (): void {
 	it('Should show warning message for SSDT project opened in Azure Data Studio', async function (): Promise<void> {
-		testContext.apiWrapper.setup(x => x.showWarningMessage(TypeMoq.It.isAny(), TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns((s) => { throw new Error(s); });
+		const stub = sinon.stub(vscode.window, 'showWarningMessage').returns(<any>Promise.resolve(constants.noString));
```

Comment:
does there need to be stub.restore() at the end of this test?

Reply:
The afterEach at the top level calls sinon.restore() which clears all wrapped functions after each test.

https://github.com/microsoft/azuredatastudio/pull/11345/files#diff-c95a71d463a38c5b2a2a1a5325b4b8cdR54